Implement `SelectMany()`

djcarter85 opened the pull request:
This is the final PR I'm expecting to need for #11 and it implements two overloads of `SelectMany()`.

A few points/questions:
1. There was a question about the behaviour of `TrySample()` when combining two distributions and the first one fails. I've decided that it's best to simply fail fast in that case, because if not we'd have to read the `result` parameter (whose value could be invalid in some way) to feed it into `selector`. What do you think about that decision? Do you think it needs additional clarity in the doc comments?
2. There was also a question about whether the implementation of `Sample()` should use the underlying implementations of `Sample()` or `TrySample()`. I've done it using `Sample()` here but let me know if you think it would be better the other way around - I _think_ it shouldn't matter but I haven't thought it through much.
3. What do you think of the test names? I'm not super keen on them as it follows a naming strategy I've just made up really. Normally I tend to write test names as a statement of fact about the behaviour of the code (something like [this blog post](https://enterprisecraftsmanship.com/posts/you-naming-tests-wrong/) advocates) but I didn't want to do that here as it didn't match what I could see in the other tests!
4. In LINQ each method is in a separate file with the class itself defined as `partial`. I wonder whether that might be a good idea here to keep related things together. Similarly we could break the tests each into separate classes or files, which might help solve the test naming problem. What do you think?

ociaw commented:
1) Yeah, I think that behavior is fine. However, it needs be explicitly documented, somewhere... Maybe just a `<remarks>` section on SelectMany?
2) Same as `TrySample`.
3) I'm not particularly attached to the existing test names - personally I still haven't found a naming scheme that I'm completely happy with. Feel free to try out a different naming scheme.
4. That's not a bad idea, it'll help keep the methods with their inner classes. Plus it makes having a bunch of inline documentation less cumbersome.

djcarter85 commented:
> 1. Yeah, I think that behavior is fine. However, it needs be explicitly documented, somewhere... Maybe just a `<remarks>` section on SelectMany?

I've added `<remarks>` section - see what you think.

> 2. Same as `TrySample`.

Not sure what you mean by this - could you elaborate?

> 3. I'm not particularly attached to the existing test names - personally I still haven't found a naming scheme that I'm completely happy with. Feel free to try out a different naming scheme.

I've tried naming the tests according to the idea of a statement of fact about the behaviour. See what you think! I'm sure the test names could be improved.

> 4. That's not a bad idea, it'll help keep the methods with their inner classes. Plus it makes having a bunch of inline documentation less cumbersome.

I've split out the prod code into partial classes, and the test code into actual classes. I think it splits things up nicely!

ociaw commented:
I think everything looks good! I've merged it all in.

(Sorry for suddenly dropping off of the face of the earth; I intend to get back to working on RandN.)
